functions/lib/github.js

```diff
@@ -1,167 +1,202 @@
-        status.target_url = 'https://sign.the.cla'
-        status.description = 'Cannot evaluate CLA for this PR. ' +
-          'Number of commits exceeds the 250 commit limit. ' +
-          'Please contact contact support@opennetworking.org'
-        return ghClient.repos.createStatus(status)
+        status.description = `We don't have a CLA in file for @${pr.user.login}`
+        status.comment = `Hi @${pr.user.login}, ` +
+          'this is the ONF bot 🤖 I\'m glad you want to contribute to ' +
+          'our projects! However, before accepting your contribution, ' +
+          'we need to ask you to sign a Contributor License Agreement ' +
+          '(CLA). You can do it online, it will take only few minutes:' +
+          '\n\n✒️ 👉 https://cla.opennetworking.org\n\n' +
+          'After signing, make sure to add your Github user ID ' +
+          `\`${pr.user.login}\` to the agreement.`
       }
+    } catch (error) {
+      console.error(error)
+      status.state = 'error'
+      status.description = 'cannot check whitelist'
+    }
 
-      // We need a CLA in file for the PR author (github ID), as well as for all
-      // the identities associated with all commits of this PR.
-      const identities = getPrIdentities(pr, ghClient)
+    // If any error occurred, improve user experience by posting a comment.
+    if (status.state === 'error' && !status.comment) {
+      status.comment =
+        `Unable to verify CLA: ${status.description}. ` +
+        'If the problem persists, please contact support@opennetworking.org ' +
+        ` (\`support-id: ${eventSnapshot.id}\`)`
+    }
 
-      return clalib.checkIdentities(identities).then(result => {
-        if (result.allWhitelisted) {
-          console.log('cla is signed for all commits')
-          status.state = 'success'
-          status.description = 'All good! We have a CLA in file for all contributors in this PR.'
+    // Post status to Github and update event in the DB
+    const statusPromise = octokit.repos
+      .createStatus({
+        owner: pr.base.repo.owner.login,
+        repo: pr.base.repo.name,
+        context: 'onf/cla',
+        sha: pr.head.sha,
+        target_url: 'https://cla.opennetworking.org',
+        description: status.description,
+        state: status.state
+      })
+      .then(() => {
+        status.githubAck = true
+      })
+      .catch(error => {
+        status.githubAck = false
+        status.githubError = JSON.parse(JSON.stringify(error))
+      })
+      .then(() => {
+        event.status = status
+        return eventSnapshot.ref.update(event)
+      })
+      .catch(console.error)
+
+    const contribRef = db.collection('contributions')
+      .doc(event.contributionId)
+    const commentPromise = contribRef.get()
+      .then(snapshot => {
+        // Retrieve existing comment_id (if any)
+        if (!snapshot.exists) {
+          console.error(`Missing contribution ${event.contributionId} in DB`)
+          return null
```

Comment:
I would `throw new Error('msg')` and handle it in the `.catch`

Reply:
In this case, I wanted to be robust against potential inconsistencies with the `contributions` collection. If the contribution doc cannot be found, it's fine to return a null comment ID. Worst case, we'll post a new comment instead of updating a new one. Using a catch would make it cleaner, but what it's not clear to me is how to distinguish between this error (contribution doc not found) and other errors. It's ok to return a null comment ID in this case, but maybe it's not ok for other cases. I guess I could define different types of errors, and take the appropriate action in the final catch... what would be the most JS-friendly approach in this case?

Reply:
The most common pattern I've seen around is to throw different errors and handle them in the final `.catch` but if that's not an error then it's fine to return `null`. In that case I would change the log to `console.warn`

This is a pretty good article on error management with promises in case you want to read about it: https://medium.com/@arthurxavier/error-handling-in-long-promise-chains-155f610b5bc6

Reply:
Thanks for the pointer. I changed the log level to warn, it makes more sense.